common/src/main/kotlin/de/hpi/cloud/common/entity/Wrapper.kt

```diff
@@ -0,0 +1,175 @@
+            val dec = decoder.beginStructure(descriptor)
+            var type: String? = null
+            var version: Int? = null
+            var id: Id<E>? = null
+            var metadata: Metadata? = null
+            var value: E? = null
+
+            loop@ while (true) {
+                when (val i = dec.decodeElementIndex(descriptor)) {
+                    CompositeDecoder.READ_DONE -> break@loop
+                    0 -> type = dec.decodeStringElement(descriptor, i)
+                    1 -> version = dec.decodeIntElement(descriptor, i)
+                    2 -> id = Id(dec.decodeStringElement(descriptor, i))
+                    3 -> metadata = dec.decodeSerializableElement(descriptor, i, Metadata.serializer())
+                    4 -> value = dec.decodeSerializableElement(descriptor, i, entitySerializer)
+                    else -> throw SerializationException("Unknown index $i")
+                }
+            }
+            dec.endStructure(descriptor)
+            return Wrapper(
+                type ?: throw MissingFieldException("type"),
+                version ?: throw MissingFieldException("version"),
+                id ?: throw MissingFieldException("id"),
+                metadata ?: throw MissingFieldException("metadata"),
+                value ?: throw MissingFieldException("value")
+            )
+        }
+    }
+
+    val documentId: String
+        get() = id.documentId(type)
+
+    // region Mutation
+    fun withValue(context: Context, newValue: E): Wrapper<E> {
+        return if (value == newValue) withKeepAlive(context)
+        else copy(
+            metadata = metadata.copy(
+                events = metadata.events + UpdateEvent.create(context, this)
+            ),
+            value = newValue
+        )
+    }
+
+    fun withKeepAlive(context: Context): Wrapper<E> {
+        return copy(
+            metadata = metadata.copy(
+                events = metadata.events + KeepAliveEvent.create(context)
+            )
+        )
+    }
+
+    fun withSources(context: Context, newSources: List<L10n<URI>>): Wrapper<E> {
+        return if (metadata.sources == newSources) this
+        else copy(
+            metadata = metadata.copy(
+                sources = newSources,
+                events = metadata.events + SourcesChangeEvent.create(context, this)
+            )
+        )
+    }
+
+    fun withPermissions(context: Context, newPermissions: Permissions): Wrapper<E> {
+        return if (metadata.permissions == newPermissions) this
+        else copy(
+            metadata = metadata.copy(
+                permissions = newPermissions,
+                events = metadata.events + PermissionsChangeEvent.create(context, this)
+            )
+        )
+    }
+
+    val isDeleted: Boolean
+        get() = metadata.eventsOfType<DeletedChangeEvent>()
+            .filter { it.isEffectiveNow }
+            .sumBy { if (it.isDeleted) -1 else 1 } < 0
```

Comment:
couldn't you iterate backwards and check for deleted events? this has to iterate the whole chain

Reply:
I added a comment explaining this behaviour: https://github.com/HPI-de/hpi-cloud/blob/58ae5203b65e3a5c17a0a557869b32cbd0dfc6de/common/src/main/kotlin/de/hpi/cloud/common/entity/Events.kt#L107-L138